src/Medidata.MAuth.Core/MAuthRequestRetrier.cs

```
@@ -0,0 +1,65 @@
+using System.Linq;
+#if !NETSTANDARD1_4
+using System.Net.Cache;
+#endif
+using System.Net.Http;
+using System.Threading.Tasks;
+
+namespace Medidata.MAuth.Core
+{
+    internal class MAuthRequestRetrier
+    {
+        private readonly HttpClient client;
+        private RetriedRequestException exception;
+
+        public MAuthRequestRetrier(MAuthOptionsBase options)
+        {
+            var signingHandler = new MAuthSigningHandler(options: new MAuthSigningOptions()
+            {
+                ApplicationUuid = options.ApplicationUuid,
+                PrivateKey = options.PrivateKey
+            },
+            innerHandler: options.MAuthServerHandler ??
+#if NETSTANDARD1_4
+                new HttpClientHandler()
+#else
+                new WebRequestHandler()
+                {
+                    CachePolicy = new RequestCachePolicy(RequestCacheLevel.Default)
+                }
+#endif
+            );
+
+            client = new HttpClient(signingHandler);
+            client.Timeout = TimeSpan.FromSeconds(options.AuthenticateRequestTimeoutSeconds);
+        }
+
+        public async Task<HttpResponseMessage> GetSuccessfulResponse(Guid applicationUuid,
+            Func<Guid, HttpRequestMessage> requestFactory, int remainingAttempts)
+        {
+            var request = requestFactory?.Invoke(applicationUuid);
+
+            if (request == null)
+                throw new ArgumentNullException(
+                    "No request function provided or the provided request function resulted null request.",
+                    nameof(requestFactory)
+                );
+
+            exception = exception ?? new RetriedRequestException(
+                $"Could not get a successful response from the MAuth Service after {remainingAttempts} attempts. " +
+                "Please see the responses for each attempt in the exception's Responses field.");
+
+            if (remainingAttempts == 0)
+                throw exception;
```

- You can throw the exception inside the `if` statement in shorter form if you want to.
`throw new RetriedRequestException() { ... }`

- There's no need for a dedicated constructor.
You already have `MAuthOptionsBase` inside `MAuthRequestRetrier`.
What you need is add a property like  
`public int NoOfAttempts { get; set; } = (int)MAuthServiceRetryPolicy + 1;`
to `MAuthOptionsBase` and use this wherever is needed.

If you want to keep this as is, that's fine.
However, is there a way not to instantiate `RetriedRequestException` on successful requests which should be 99% of the time.